### PR TITLE
clprover: move to by-name, windows and aarch compatability

### PR DESCRIPTION
**Pull request by mgttlinger**
## Things done

Moved the `clprover` package to the `by-name` hierarchy and included windows into the build platforms. It currently fails to cross-compile (`mingwW64`) due to missing `g++` compiler but adding a `gcc` dependency doesn't seem to fix it as this is not available for that platform. I'm not experienced with targeting Windows so I keep it marked as broken for now.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/cl/clprover/package.nix [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    install -m755 bin/${pname} $out/bin/${pname}
+
+    mkdir -p $out/share/${pname}
+    cp -r examples $out/share/${pname}/examples
+
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    description = "An ordered resolution based theorem prover for Coalition Logic.";
```
Comment (FliegendeWurst) [inline pkgs/by-name/cl/clprover/package.nix:43]:
```suggestion
    description = "Ordered resolution based theorem prover for Coalition Logic";
```

File: pkgs/by-name/cl/clprover/package.nix [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    mkdir -p $out/bin
+    install -m755 bin/${pname} $out/bin/${pname}
+
+    mkdir -p $out/share/${pname}
+    cp -r examples $out/share/${pname}/examples
+
+    runHook postInstall
+  '';
+
+  meta = with lib; {
```
Comment (FliegendeWurst) [inline pkgs/by-name/cl/clprover/package.nix:42]:
```suggestion
  meta = {
```

File: pkgs/by-name/cl/clprover/package.nix [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    description = "An ordered resolution based theorem prover for Coalition Logic.";
+    homepage = "https://cgi.csc.liv.ac.uk/~ullrich/CLProver++/";
+    maintainers = with maintainers; [ mgttlinger ];
+    platforms = with platforms; linux ++ windows;
+    broken = stdenv.hostPlatform.isWindows; # g++ is missing there in my tests but a gcc input is also not available there.
+    license = licenses.gpl3;
```
Comment (FliegendeWurst) [inline pkgs/by-name/cl/clprover/package.nix:48]:
```suggestion
    maintainers = with lib.maintainers; [ mgttlinger ];
    platforms = with lib.platforms; linux ++ windows;
    broken = stdenv.hostPlatform.isWindows; # g++ is missing there in my tests but a gcc input is also not available there.
    license = lib.licenses.gpl3;
```

File: pkgs/by-name/cl/clprover/package.nix [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+    sed -i 's/\(TARGET_OS *:= *\)LINUX/\1WINDOWS/g' Makefile
+
+    runHook postConfigure
+  '';
+
+  preBuild = ''
+    mkdir bin
+  '';
+
```
Comment (FliegendeWurst) [inline pkgs/by-name/cl/clprover/package.nix:29]:
```suggestion

  makeFlags = [
    "CC=${lib.getBin stdenv.cc}/bin/${stdenv.cc.targetPrefix}cc"
  ];
```
This partially fixes cross-compiling.

**Review by FliegendeWurst — CHANGES_REQUESTED** [review]

**Comment by Pandapip1** [comment]
Merge conflict due to https://github.com/NixOS/nixpkgs/pull/354531.

**Comment by mgttlinger** [comment]
Thanks for all the feedback and I got cross compilation to work fully.

File: pkgs/by-name/cl/clprover/package.nix [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    cp -r examples $out/share/${self.pname}/examples
+
+    runHook postInstall
+  '';
+
+  meta = {
+    description = "Ordered resolution based theorem prover for Coalition Logic";
+    homepage = "https://cgi.csc.liv.ac.uk/~ullrich/CLProver++/";
+    maintainers = with lib.maintainers; [ mgttlinger ];
+    platforms = with lib.platforms; linux ++ windows;
```
[the remark this answers is not in the document]
Comment (mgttlinger) [reply pkgs/by-name/cl/clprover/package.nix:62]: The build got ever more complicated, but now works for 32-bit architectures and aarch.

File: pkgs/by-name/cl/clprover/package.nix [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    sed -i 's/-m64/${
+      if stdenv.hostPlatform.isArmv7 then
+        "-march=armv7"
+      else if stdenv.hostPlatform.isAarch64 then
+        "-march=armv8"
+      else if stdenv.hostPlatform.is32bit then
+        "-m32"
+      else
+        "-m64"
+    }/g' Makefile
```
Comment (Pandapip1) [inline pkgs/by-name/cl/clprover/package.nix:33]: This should be modifying the Makefile in postPatch

File: pkgs/by-name/cl/clprover/package.nix [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    cp -r examples $out/share/${self.pname}/examples
+
+    runHook postInstall
+  '';
+
+  meta = {
+    description = "Ordered resolution based theorem prover for Coalition Logic";
+    homepage = "https://cgi.csc.liv.ac.uk/~ullrich/CLProver++/";
+    maintainers = with lib.maintainers; [ mgttlinger ];
+    platforms = with lib.platforms; linux ++ windows;
```
Comment (Pandapip1) [inline pkgs/by-name/cl/clprover/package.nix:62]: Nit: This is the more commonly used formt

```suggestion
    platforms = with lib.platforms; [
      linux
      windows
    ];
```

Comment (mgttlinger) [reply pkgs/by-name/cl/clprover/package.nix:58]: That doesn't type as both `linux` and `windows` are lists of platform strings.

Comment (Pandapip1) [reply pkgs/by-name/cl/clprover/package.nix:58]: Please disregard and mark as resolved

**Review by FliegendeWurst — APPROVED** [review]

**Review by Pandapip1 — APPROVED** [review]

**Comment by FliegendeWurst** [comment]
aarch64-linux fails, other than that it looks good. Unfortunately ofborg is somewhat slow, so I often don't see the build results when checking on the PR again.

```
cc1plus: error: unknown value 'armv8' for '-march'
cc1plus: note: valid arguments are: armv8-a armv8.1-a armv8.2-a armv8.3-a armv8.4-a armv8.5-a armv8.6-a armv8.7-a armv8.8-a armv8-r armv9-a armv9.1-a armv9.2-a armv9.3-a native; did you mean 'armv8-a'?
```

**Review by FliegendeWurst — APPROVED** [review]
`{,pkgsCross.aarch64-multiplatform.,pkgsCross.mingwW64.,pkgsCross.mingw32.}clprover` all build here on x64 Linux.

**Pull request merged**